Fix py37-38 structuring of mixed typing/regular generics

bibajz opened the pull request:
Fixes #218

(EDIT 3.2.) I added tests describing the following behaviours:

* attribute annotated by a bare lower-case generic - `dict`, `frozenset`, `list`, `set`, `tuple`
* attribute annotated by a `typing.` upper-case generic with 1 type parameter, which can itself be a lower-case generic - this excludes sets and tuple
* attribute annotated by `typing.Tuple` with variadic lower-case generics


Cheers,
Libor

bibajz commented on src/cattr/_compat.py:
```diff
@@ -161,21 +161,11 @@ def is_mapping(type):
-            or args == bare_mutable_seq_args
-        )
+        # Lower-cased generics in 3.7-8 do not have `__args__` attribute.
+        # Generics imported from `typing` module which do not have their
```
This is incorrect and now, although the reported bug is fixed, everything else is broken lol.

Sorry, brain fart, will fix it soon.

Tinche commented:
Is this ready for review?

bibajz commented:
Hi @Tinche, made some dumb mistakes prior, but now, it is ready! :)

Tinche commented:
The implementation looks good, it needs a changelog entry and possibly documentation updates.

We already have a lot of Hypothesis tests though, I don't want to add new ones. What we should do is change the existing tests to test for these. I will push a branch with some modified tests you can rebase on and see what I mean.

bibajz commented:
OK, I can make a changelog entry.

Regarding the hypothesis tests, would it make sense to hide my added tests behind conditional skip - `is_py38 or is_py37`?

Tinche commented:
Here's the new branch: https://github.com/python-attrs/cattrs/tree/tin/new-tests

If you look at the last commit, I added bare `list`, `dict`, `set` and `frozenset` to the strategies and enabled from for 3.7/3.8. So the strategies are already parametrized ;)

bibajz commented:
And also, that is for a separate issue, but docs for contributing are outdated - poetry does not support editable installs - 

```
$ pip install -e.
Obtaining file:///home/libor/Desktop/Programovani/Pythony/forked/cattrs
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
ERROR: Project file:///home/libor/Desktop/Programovani/Pythony/forked/cattrs has a 'pyproject.toml' and its build backend is missing the 'build_editable' hook. Since it does not have a 'setup.py' nor a 'setup.cfg', it cannot be installed in editable mode. Consider using a build backend that supports PEP 660.
```

Tinche commented:
Ah, we were using the alpha release of Poetry 1.2 for a while, which does support this, but fell back to 1.1 for other reasons.

bibajz commented:
Ok, checked the branch you posted, I will

1. revert all the changes in `tests/` directory on this branch
2. cherry-pick your commit
3. add changelog entry
4. push and profit

Tinche commented:
Sounds good. I think I missed a few, like `tuple`?

bibajz commented:
Ah, true, so will in the types you missed and $$$

bibajz commented:
Hey @Tinche, it's ready for review - 

I have added changelog entry, tests for `tuple`s and also did some clean-up of `collections.abc.` aliases that were there for some reason, but I do not understand why exactly, when you can import them from `typing`.

Did I miss something?

bibajz commented on HISTORY.rst:
```diff
@@ -7,6 +7,8 @@ History
 * ``attrs`` and dataclass structuring is now ~25% faster.
 * Fix an issue structuring bare ``typing.List`` s on Pythons lower than 3.9.
   (`#209 <https://github.com/python-attrs/cattrs/issues/209>`_)
+* Fix structuring bare generics on Pythons lower than 3.9.
```
Maybe be more specific with `bare lower-case generics`

Tinche replied:
The thing is, these aren't even generics on 3.7. So more like `non-parameterized containers like list and dict` I guess?

bibajz replied:
Yep, will do!

bibajz commented:
Well, it seems that tests uncovered another failing case: 

```python
>>> import attrs
>>> import cattrs
>>> from typing import Tuple

>>> @attrs.define
... class A:
...     a: Tuple = attrs.field(default=('', '', ''))

>>> cattrs.unstructure(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/converters.py", line 197, in unstructure
    return self._unstructure_func.dispatch(
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/dispatch.py", line 49, in _dispatch
    return self._function_dispatch.dispatch(cl)
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/dispatch.py", line 126, in dispatch
    return handler(typ)
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/converters.py", line 722, in gen_unstructure_attrs_fromdict
    h = make_dict_unstructure_fn(
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/gen.py", line 125, in make_dict_unstructure_fn
    handler = converter._unstructure_func.dispatch(t)
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/dispatch.py", line 49, in _dispatch
    return self._function_dispatch.dispatch(cl)
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/dispatch.py", line 126, in dispatch
    return handler(typ)
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/converters.py", line 754, in gen_unstructure_iterable
    h = make_iterable_unstructure_fn(
  File "/home/libor/.virtualenvs/p38/lib/python3.8/site-packages/cattr/gen.py", line 339, in make_iterable_unstructure_fn
    type_arg = get_args(cl)[0]
IndexError: tuple index out of range
```

bibajz commented:
Again, failed on py37/8, is OK on py39 (and pressumably py310)

bibajz commented:
I have pushed 2 commits

 * since I discovered that `Tuple.__args__ == ()`, I added this check to `is_bare`.
 * add the same check in unstructuring iterable, because tests were failing on unstructuring unannotated `Tuple`s

What remains to be fixed are tests for roundtrips of un/struct of sets and frozen sets.

I am out for today, will investigate later.

Cheers, Libor

bibajz commented:
Also, all the checks around the code for `if getattr(..., '__args__', None) ,,,` seem to be very fragile and I suggest to consolidate them in one place - I understand it's a pain in the ass, since 4 versions of Python are supported and typing in Python has undergone some huge development,  but it seems to be quite a source of bugs here.   :-/

Tinche commented:
I'm gonna have to ask you to return the `abc` generics. On newer Pythons, classes like `typing.MutableSequence` are deprecated in favor of `collections.abc.MutableSequence`, so we need to test them too.

(https://docs.python.org/3/library/typing.html#typing.MutableSequence)

bibajz commented:
> I'm gonna have to ask you to return the `abc` generics. On newer Pythons, classes like `typing.MutableSequence` are deprecated in favor of `collections.abc.MutableSequence`, so we need to test them too.
> 
> (https://docs.python.org/3/library/typing.html#typing.MutableSequence)

Yes, got it, that's what I missed from the documentation.

bibajz commented on src/cattrs/gen.py:
```diff
@@ -412,8 +412,10 @@ def make_iterable_unstructure_fn(cl: Any, converter, unstructure_to=None):
+    # Let's try fishing out the type args
+    # Unspecified tuples have `__args__` as empty tuples, so guard
+    # against IndexError.
+    if getattr(cl, "__args__", None) not in (None, ()):
```
Btw @Tinche, I could cherry-pick this commit and create a separate branch, because this is related to #209 .

That fix was not only for `t.List`, but also for `t.Set`, `t.FrozenSet` etc., but `t.Tuple` has, as I noted earlier, `t.Tuple.__args__ == ()`

Tinche replied:
I wouldn't bother with another branch

bibajz commented:
OK, now, it should be complete:

* I have refactored the `is_bare`  check as a set membership
* Added a changelog mentioning the `Tuple` structuring
* Finished the `tests/metadata/__init__.py`

What do you think? @Tinche

Tinche commented:
Looks great! I think the changelog format needs to be tweaked but I can do that myself. Thanks!
